modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Resources.scala

```diff
@@ -116,45 +127,112 @@ object Resources {
     */
   private def mkGoodSink[F[_]: Concurrent: ContextShift: Timer](
     blocker: Blocker,
     good: Output.BigQuery,
     bigQuery: BigQuery,
     producer: PubsubProducer[F, FailedInsert],
     metrics: Metrics[F],
-    typesSink: Pipe[F, Set[ShreddedType], Unit]
+    typeSink: Pipe[F, Set[ShreddedType], Unit],
+    bufferOverflowQueue: Queue[F, StreamLoaderRow[F]],
+    sinkSettingsGood: SinkSettings.Good,
+    sinkSettingsTypes: SinkSettings.Types
   ): Pipe[F, StreamLoaderRow[F], Unit] = {
-    val goodPipe: Pipe[F, StreamLoaderRow[F], Unit] = _.parEvalMapUnordered(MaxConcurrency) { slrow =>
-      blocker.blockOn(
-        Bigquery.insert(producer, metrics, slrow.row)(Bigquery.mkInsert(good, bigQuery)) *> slrow.ack
-      )
+    val goodPipe: Pipe[F, StreamLoaderRow[F], Unit] = {
+      _.merge(bufferOverflowQueue.dequeue)
+        .through(batch(sinkSettingsGood.bqWriteRequestThreshold, sinkSettingsGood.bqWriteRequestTimeout))
+        .parEvalMapUnordered(sinkSettingsGood.sinkConcurrency) { slrows =>
+          val (remaining, toInsert) = splitBy(slrows, getSize[F], sinkSettingsGood.bqWriteRequestSizeLimit)
+          remaining.traverse_(bufferOverflowQueue.enqueue1) *>
+            blocker.blockOn(
+              Bigquery.insert(producer, metrics, toInsert.map(_.row))(Bigquery.mkInsert(good, bigQuery)) *> toInsert
+                .traverse_(_.ack)
+            )
+        }
+    }
+
+    _.observe(goodPipe)
+      .map(_.row.inventory)
+      .through(aggregateTypes[F](sinkSettingsTypes.batchThreshold, sinkSettingsTypes.batchTimeout))
+      .through(typeSink)
+  }
+
+  /**
+    * Splits a List of As into two Lists, using the provided getSize function.
+    * Used to split a List[StreamLoaderRow] into a (remaining, toInsert) Lists.
+    * @param batch A List of As (used with StreamLoaderRow)
+    * @param getSize A function to get the size of a single A
+    * @param threshold A limit to how big the toInsert List can be, as measured by toInsert.map(getSize).sum
+    * @tparam A The type of element in the original collection (used with StreamLoaderRow)
+    * @return A tuple of Lists. The second List contains all As that could be inserted into it without breaching the threshold. The first List contains the remaining As.
+    */
+  private[streamloader] def splitBy[A](batch: List[A], getSize: A => Int, threshold: Int): (List[A], List[A]) = {
+    val originalBatch = batch
+    @tailrec
+    def go(batch: List[A], acc: (List[A], Int, Boolean)): (List[A], List[A]) = {
+      val (buffer, totalSize, isFull) = acc
+      batch match {
+        case _ if isFull => (originalBatch.drop(buffer.length), buffer.reverse)
+        case Nil         => (List.empty[A], buffer.reverse)
+        case h :: t =>
+          val currSize = getSize(h)
+          if (currSize + totalSize > threshold) go(t, (buffer, totalSize, true))
+          else go(t, (h :: buffer, currSize + totalSize, false))
+      }
     }
```

Comment:
This seems very complicated for a simple problem. You could just go through the list until the sum of elements reach `threshold` and then use `splitAt` at the index.

Reply:
Hey Ben. I can't think of how this would significantly simplify things. We'll still need to keep a running total size, so we can compare it with the threshold; we'll have to also keep track of the element's index; and `splitAt` still needs to go through the full List.

I'm very happy to simplify it but I'm just not seeing how. Any suggestions welcome.

Reply:
I was thinking something like

```
val (_, index) = l.foldLeft((0, 0)){case (acc, elem) => if (acc._1 + elem >= threshold) acc else (acc._1 + elem, acc._2 +1)}
l.splitAt(index)
```

Reply:
Ah, I see. But wouldn't `foldLeft` process the entire list, even after it's clear that the threshold has been reached?

Reply:
Yes it would. I missed the fact that your function was not and just wanted to provide a more concise implementation, sorry